### DISCUSSION TITLE
Stock prompt from Xonsh

### DISCUSSION THREAD

**Discussion by rickysarraf** [xonsh/xonsh#5453] [Q&A]
First, thank you so much for Xonsh. It is taking a while in the busy schedule, but once relearnt; this is quite enjoyable.

So far, my biggest frustration is the Xonsh Prompt setup. From what it turns out, there are quite a limited number of xontribs available; and of them not everything is maintained.

Then, there's also certain stock features missing in these 3rd party prompt xontribs. Like, for example, xontrib-powerline2 looks decent but doesn't support all attributes of `$PROMPT_FIELDS` like `gitstatus`. `powerline3` claims to support everything but looks not good on my setup.

I would humbly request the upstream project to create one stock PROMPT; like the current one with PROMPT_TOOLKIT. And please test all old/new features and functionalities against that PROMPT; as an exit criteria for a release.

Right now, with Xonsh 0.16, I have a broken BOTTOM_TOOLBAR with stock Xonsh prompt and I have spent more than my allowed time to figure out just the prompt; something which I could use in learning the cool new features of Xonsh.

![image](https://github.com/xonsh/xonsh/assets/65480/b4ddd9ae-1578-442e-9154-87d66b3c2f72)
Any kind soul that can help explain why the bottom bar has this large ugly white patch.

And my stock_prompt() definition being:

```python
def stock_prompt():
  $DYNAMIC_CWD_WIDTH="20"
  #$DYNAMIC_CWD_ELISION_CHAR="..."
  $PROMPT = '{env_name}{BOLD_GREEN}{user}@{hostname}{BOLD_BLUE} {cwd} {RED}{last_return_code_if_nonzero:[{BOLD_INTENSE_RED}{}{RED}] }{RESET}{BOLD_BLUE}{prompt_end}{RESET} '
  $BOTTOM_TOOLBAR='{RESET}{DEFAULT}{aquarius_signs} {localtime} {last_return_code_if_nonzero} {cwd_dir} {env_name} {gitstatus} {current_job}{DEFAULT}'
  $RIGHT_PROMPT='{last_return_code_if_nonzero} {localtime}'
  print("Loaded function prompt: ", (inspect.currentframe().f_code.co_name))
stock_prompt()
```

My setup:

```
rrs@xps ~ @ xonfig   
+------------------+--------------------+
| xonsh            | 0.16.0             |
| Python           | 3.11.9             |
| PLY              | 3.11               |
| have readline    | True               |
| prompt toolkit   | 3.0.44             |
| shell type       | prompt_toolkit     |
| history backend  | json               |
| pygments         | 2.18.0             |
| on posix         | True               |
| on linux         | True               |
| distro           | debian             |
| on wsl           | False              |
| on darwin        | False              |
| on windows       | False              |
| on cygwin        | False              |
| on msys2         | False              |
| is superuser     | False              |
| default encoding | utf-8              |
| xonsh encoding   | utf-8              |
| encoding errors  | surrogateescape    |
| xontrib 1        | coreutils          |
| xontrib 2        | histcpy            |
| xontrib 3        | autojump           |
| xontrib 4        | readable-traceback |
| xontrib 5        | sh                 |
| xontrib 6        | hist_navigator     |
| xontrib 7        | abbrevs            |
| xontrib 8        | output_search      |
| RC file 1        | /home/rrs/.xonshrc |
+------------------+--------------------+
rrs@xps ~ @         
```

**Comment by anki-code** [comment]
Hey @rickysarraf! Thank you for reporting this!
I would recommend [xonsh-cheatsheet](https://github.com/anki-code/xonsh-cheatsheet/blob/main/README.md) to you for further diving into xonsh.
Then take a look into [xontrib-prompt-bar](https://github.com/anki-code/xontrib-prompt-bar) and [xontrib-prompt-starship](https://github.com/anki-code/xontrib-prompt-starship/).
I'm not remember the white line when I tested bottom bar with starship.

**Reply by rickysarraf** [reply]
Thank you for replying back @anki-code

I have been using a great deal of work that you have done. So being able to thank in person is great 🙏🏽 

I tried them both as well. They are decent but not comprehensive. And they are all 3rd party xontribs.

My request is more to do with a stock recommended prompt by the project. One that is well tested with every release of Xonsh. 



Here's what `prompt_bar` looks like. There was no bug free consistency.

For example, I have;

![image](https://github.com/xonsh/xonsh/assets/65480/f933e2a8-8d35-49af-a927-278d1813ea9b)


with below code:

```python
def prompt_bar():

  $XONTRIB_PROMPT_BAR_WRAPPERS = {
    'brackets': lambda v: f'[{v}]'
  }

  $XONSH_COLOR_STYLE = "paraiso-dark"
  #$XONSH_COLOR_STYLE="default"
  $XONTRIB_PROMPT_BAR_LEFT = '{hostname}{user}{cwd_abs#accent}{gitstatus#section}'
  $XONTRIB_PROMPT_BAR_RIGHT = '{aquarius_signs#section#brackets}{last_return_code_if_nonzero#accent#brackets}{hist_status}{screens#section}{env_name#strip_brackets#section}{date_time_tz}'

  # $XONTRIB_PROMPT_BAR_THEME = {
  #     'left':       '{hostname}{user}{cwd_abs}',
  #     'right':      '{curr_branch}{env_name}{date_time_tz}',
  #     'bar_bg':     '{BACKGROUND_#808080}',
  #     'bar_fg':     '{#AAA}',
  #     'section_bg': '{BACKGROUND_#444666}',
  #     'section_fg': '{#CCC}',
  #     'accent_fg':  '{BOLD_#DDD}',
  # }

  xontrib load cmd_done
  xontrib load prompt_bar
  print("Loaded function prompt: ", (inspect.currentframe().f_code.co_name))
prompt_bar()
```

And I have no clue why `prompt_bar` will not honor `#section` field; or for that matter any of those `#fields`, in 90 out of the 100 invocations of Xonsh. In 10 random invocations, it does get the color code right.




@anki-code Please see the inconsistency in between 2 invocations of Xonsh on the same terminal in different tabs.

![image](https://github.com/xonsh/xonsh/assets/65480/a2429b18-ee54-44f3-b5be-4446bd1f879b)
The above image with a light bar for the prompt. And on top of that each `#section` item is rendered correct.


![image](https://github.com/xonsh/xonsh/assets/65480/d56a350c-a50f-46c9-8e64-e73464529489)
Now, this very same software configuration, just running in another tab, with not identical results.

**Reply by anki-code** [reply]
It will be cool to repeat bottom toolbar white line on pure prompt-toolkit - example - https://github.com/prompt-toolkit/python-prompt-toolkit/blob/6a58564f6a201f1234733f726d866a64e95b6ba3/examples/prompts/bottom-toolbar.py
Just run it

**Reply by anki-code** [reply]
I am very surprised by your experience. I used prompt_bar on Windows WSL with Windows Terminal, on Manjaro Linux with KDE Konsole, on Mac with iTerm and every time it looks like this:

![image](https://github.com/xonsh/xonsh/assets/1708680/3856115a-44b2-4321-9bf3-9627e8ab029b)

Why you lost colors it's the good question. On the first looks it pretty depends on terminal app settings.

**Reply by rickysarraf** [reply]
> It will be cool to repeat bottom toolbar white line on pure prompt-toolkit - example - https://github.com/prompt-toolkit/python-prompt-toolkit/blob/6a58564f6a201f1234733f726d866a64e95b6ba3/examples/prompts/bottom-toolbar.py Just run it

Thank you. This is the result I have
![image](https://github.com/xonsh/xonsh/assets/65480/ce4d5fc1-62df-410d-8550-cf1f288f2736)

**Reply by rickysarraf** [reply]
> I am very surprised by your experience. I used prompt_bar on Windows WSL with Windows Terminal, on Manjaro Linux with KDE Konsole, on Mac with iTerm and every time it looks like this:
> 
> Why you lost colors it's the good question. On the first looks it pretty depends on terminal app settings.

I have tried it on 2 terminal applications; KDE's Konsole and Kitty. Both exhibit the same erratic behavior.

**Reply by anki-code** [reply]
It looks like the toolbar highlighting with background color is the prompt-toolkit default. I believe the default can be replaced.
